File: backend/src/modules/admin/admin.controller.ts
import {
  Controller,
  Get,
  Put,
  Delete,
  Param,
  Body,
  Query,
  UseGuards,
  HttpCode,
  HttpStatus,
  ParseIntPipe,
  DefaultValuePipe,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiCookieAuth,
  ApiQuery,
} from '@nestjs/swagger';
import { BiodataService } from '../biodata/biodata.service';
import { UpdateBiodataDto } from '../biodata/dto/biodata.dto';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import { RolesGuard } from '../../common/guards/roles.guard';
import { Roles } from '../../common/decorators/roles.decorator';
import { Role } from '../../common/enums/role.enum';

@ApiTags('Admin')
@ApiCookieAuth('access_token')
@UseGuards(JwtAuthGuard, RolesGuard)
@Roles(Role.ADMIN)
@Controller('admin')
export class AdminController {
  constructor(private readonly biodataService: BiodataService) {}

  @Get('biodata')
  @ApiOperation({ summary: 'Get all biodata (admin only)' })
  @ApiQuery({ name: 'page', required: false, type: Number })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @ApiQuery({ name: 'search', required: false })
  @ApiQuery({ name: 'appliedPosition', required: false })
  @ApiQuery({ name: 'latestEducation', required: false })
  async getAllBiodata(
    @Query('page', new DefaultValuePipe(1), ParseIntPipe) page: number,
    @Query('limit', new DefaultValuePipe(10), ParseIntPipe) limit: number,
    @Query('search') search?: string,
    @Query('appliedPosition') appliedPosition?: string,
    @Query('latestEducation') latestEducation?: string,
  ) {
    const result = await this.biodataService.getAdminBiodataList({
      page,
      limit,
      search,
      appliedPosition,
      latestEducation,
    });
    return {
      success: true,
      message: 'Data biodata berhasil diambil',
      data: result.data,
      meta: result.meta,
    };
  }

  @Get('biodata/:id')
  @ApiOperation({ summary: 'Get biodata detail (admin only)' })
  async getBiodataById(@Param('id') id: string) {
    const data = await this.biodataService.getAdminBiodataById(id);
    return { success: true, message: 'Detail biodata berhasil diambil', data };
  }

  @Get('jobs/:jobId')
  @ApiOperation({ summary: 'Poll admin job status' })
  async getJobStatus(@Param('jobId') jobId: string) {
    const data = await this.biodataService.getJobStatus(jobId);
    return { success: true, message: 'Status job berhasil diambil', data };
  }

  @Put('biodata/:id')
  @HttpCode(HttpStatus.ACCEPTED)
  @ApiOperation({ summary: 'Enqueue biodata update by id (admin, async)' })
  async updateBiodata(@Param('id') id: string, @Body() dto: UpdateBiodataDto) {
    const jobId = await this.biodataService.enqueueAdminUpdate(id, dto);
    return {
      success: true,
      message: 'Permintaan perubahan biodata sedang diproses',
      data: { jobId },
    };
  }

  @Delete('biodata/:id')
  @HttpCode(HttpStatus.ACCEPTED)
  @ApiOperation({ summary: 'Enqueue biodata deletion by id (admin, async)' })
  async deleteBiodata(@Param('id') id: string) {
    const jobId = await this.biodataService.enqueueAdminDelete(id);
    return {
      success: true,
      message: 'Permintaan penghapusan biodata sedang diproses',
      data: { jobId },
    };
  }
}

File: BackendNet/Controllers/AdminController.cs
using BackendNet.Application.Common;
using BackendNet.Application.DTOs;
using BackendNet.Application.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace BackendNet.Controllers;

[ApiController]
[Route("api/admin")]
[Authorize(Roles = "ADMIN")]
[Produces("application/json")]
public class AdminController(IBiodataService svc) : ControllerBase
{
    // GET /api/admin/biodata
    [HttpGet("biodata")]
    public async Task<IActionResult> GetAllBiodata(
        [FromQuery] int page = 1,
        [FromQuery] int limit = 10,
        [FromQuery] string? search = null,
        [FromQuery] string? appliedPosition = null,
        [FromQuery] string? latestEducation = null)
    {
        var (data, meta) = await svc.GetAdminBiodataListAsync(
            page, limit, search, appliedPosition, latestEducation);

        // Return with meta at top level (matching NestJS response shape)
        return Ok(new
        {
            success = true,
            message = "Data biodata berhasil diambil",
            data,
            meta,
        });
    }

    // GET /api/admin/biodata/:id
    [HttpGet("biodata/{id}")]
    public async Task<IActionResult> GetBiodataById(string id)
    {
        var data = await svc.GetAdminBiodataByIdAsync(id);
        if (data is null) return NotFound(new ApiErrorResponse(false, "Biodata tidak ditemukan"));
        return Ok(ApiResponse<BiodataResponse>.Ok(data, "Detail biodata berhasil diambil"));
    }

    // GET /api/admin/jobs/:jobId
    [HttpGet("jobs/{jobId}")]
    public async Task<IActionResult> GetJobStatus(string jobId)
    {
        var data = await svc.GetJobStatusAsync(jobId);
        return Ok(ApiResponse<JobStatusResponse>.Ok(data, "Status job berhasil diambil"));
    }

    // PUT /api/admin/biodata/:id  — returns 202
    [HttpPut("biodata/{id}")]
    public IActionResult UpdateBiodata(string id, [FromBody] BiodataRequest req)
    {
        var jobId = svc.EnqueueAdminUpdate(id, req);
        return Accepted(ApiResponse<JobEnqueuedResponse>.Ok(
            new JobEnqueuedResponse(jobId),
            "Permintaan perubahan biodata sedang diproses"));
    }

    // DELETE /api/admin/biodata/:id  — returns 202
    [HttpDelete("biodata/{id}")]
    public IActionResult DeleteBiodata(string id)
    {
        var jobId = svc.EnqueueAdminDelete(id);
        return Accepted(ApiResponse<JobEnqueuedResponse>.Ok(
            new JobEnqueuedResponse(jobId),
            "Permintaan penghapusan biodata sedang diproses"));
    }
}
